File: LeetCode/Hard/1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class NodeValue {
public:
    int minNode, maxNode, sumBSTs;

    NodeValue(int sum, int mn, int mx) {
        sumBSTs = sum;
        minNode = mn;
        maxNode = mx;
    }
};
class Solution {
public:
    NodeValue helper(int& res, TreeNode* node) {
        if (node == NULL)
            return NodeValue(0, INT_MAX, INT_MIN);

        NodeValue left = helper(res, node->left);
        NodeValue right = helper(res, node->right);

        if (left.maxNode < node->val && node->val < right.minNode) {
            int cur_sum = node->val + left.sumBSTs + right.sumBSTs;
            int cur_min_value = min(left.minNode, node->val); 
            int cur_max_value = max(right.maxNode, node->val);
            
            res = max(res, cur_sum);
            return NodeValue(cur_sum, cur_min_value, cur_max_value);
        }

        return NodeValue(max(left.sumBSTs, right.sumBSTs), INT_MIN, INT_MAX);
    }
    int maxSumBST(TreeNode* root) { 
        int res = 0;
        helper(res, root);  
        return res;
    }
};

File: LeetCode/Hard/1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxSum;
    private static class SubtreeInfo {
        public boolean isBST;
        public int minVal;
        public int maxVal;
        public int sum;

        public SubtreeInfo(boolean isBST, int minVal, int maxVal, int sum) {
            this.isBST = isBST;
            this.minVal = minVal;
            this.maxVal = maxVal;
            this.sum = sum;
        }
    }

    private SubtreeInfo dfs(TreeNode node) {
        if(node == null) return new SubtreeInfo(true, Integer.MAX_VALUE, Integer.MIN_VALUE, 0);

        SubtreeInfo left = dfs(node.left);
        SubtreeInfo right = dfs(node.right);

        if((left.isBST && right.isBST) && (left.maxVal < node.val && right.minVal > node.val)) {
            int sum = node.val + left.sum + right.sum;
            int currMin = Math.min(node.val, left.minVal);
            int currMax = Math.max(node.val, right.maxVal);

            maxSum = Math.max(maxSum, sum);
            return new SubtreeInfo(true, currMin, currMax, sum);
        }

        return new SubtreeInfo(false, 0, 0, 0);
    }

    public int maxSumBST(TreeNode root) {
        maxSum = 0;
        dfs(root);
        return maxSum;
    }
}